### PR TITLE
Update find to become a filter

### LINKED ISSUE #101: Update find to become a filter
can do find KEYWORD
or find role/<role> nric/<nric> etc

for now we can implement only role, nric and tags

link to #103

### PR DESCRIPTION
**Pull request by suhayl13**
### Description of the New Find Feature

**Summary:**
This pull request introduces an enhanced `find` command that allows users to locate persons in the address book based on additional criteria such as name, NRIC, role, and tags. The goal of this feature is to provide more flexibility and precision when searching for contacts, making it easier for users to find specific records based on their needs.

**Key Updates:**
- **Enhanced `find` Command**: The `find` command can now accept multiple prefixes to search for persons based on the following fields:
  - `n/NAME`: Search by name (e.g., `find n/John`).
  - `nric/NRIC`: Search by NRIC (e.g., `find nric/S1234567A`).
  - `role/ROLE`: Search by role, either `patient` or `caregiver` (e.g., `find role/patient`).
  - `t/TAG`: Search by tag(s) associated with persons (e.g., `find t/friend`).

- **Case-Insensitive Search**: All search criteria are case-insensitive, allowing `find n/alex` to match `Alex`, `ALEX`, or `alex`.

- **Support for Multiple Criteria**: The `find` command can now accept multiple criteria in a single search query. For example, `find n/Alex nric/S1234567A role/patient t/friend` will return all persons who match any of the specified criteria.

- **Preamble Handling**: The command now handles preambles correctly, ensuring that only valid criteria are considered.  For example, `find 1 n/Alex` would raise an error as indexes are not a valid argument here.

**Example Usage:**
- `find n/John nric/S1234567A`: Returns persons named `John` or with the NRIC `S1234567A`.
- `find role/patient t/friend`: Returns all `patients` or persons tagged as `friend`.

**Important!**
- The `findnric` command is now depreciated as this new find feature can perform nric search as well.
- As a result, all classes that were created for `findnric` command (including test files) have been removed.
- `NameContainsKeywordsPredicate` has been refactored into `ContainsKeywordPredicate` since multiple search criterias are now supported with the new `find` command and not just the name.

**Notes for Reviewers:**
- Please review the updated `find` command parser to ensure that the handling of prefixes and argument mappings is clear and robust.
- Pay special attention to the `ContainsKeywordsPredicate` and its handling of various search criteria.
- I’ve added test cases for each new search field and combination of fields; let me know if there are additional scenarios you think we should cover.
- The `toString` method has been updated to handle dynamic fields, making it adaptable to various search criteria. Please verify that it meets our expected output for logging and debugging.

**Why This Feature is Useful:**
This feature allows users to conduct more precise searches in the address book, which is especially useful when managing a large number of contacts with different roles or tags. It aligns with the needs of independent geriatricians to filter records based on specific criteria, improving workflow efficiency.

Fixes #101.

### CODE REVIEW

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 1]
```diff
@@ -15,14 +15,14 @@ public class FindCommand extends Command {
 
     public static final String COMMAND_WORD = "find";
 
-    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all persons whose names contain any of "
-            + "the specified keywords (case-insensitive) and displays them as a list with index numbers.\n"
-            + "Parameters: KEYWORD [MORE_KEYWORDS]...\n"
-            + "Example: " + COMMAND_WORD + " alice bob charlie";
+    public static final String MESSAGE_USAGE = COMMAND_WORD + ": Finds all persons based on the specified criteria "
```
Comment (Ponharish) [inline src/main/java/seedu/address/logic/commands/FindCommand.java:18]: Am I right to assume that not all of the parameters such as name, nric, role, tags are required?
In such a case would it be better to state that not all parameters (such as name, nric, role, ...) are required by the function?

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java [hunk 2]
```diff
@@ -15,7 +15,6 @@
 import seedu.address.logic.commands.EditCommand;
 import seedu.address.logic.commands.ExitCommand;
 import seedu.address.logic.commands.FindCommand;
-import seedu.address.logic.commands.FindNricCommand;
```
Comment (Ponharish) [inline src/main/java/seedu/address/logic/parser/AddressBookParser.java:18]: It is good that you have removed all traces of findnric command and hence avoided dead code

File: src/main/java/seedu/address/logic/commands/FindCommand.java [hunk 3]
```diff
@@ -15,14 +15,14 @@ public class FindCommand extends Command {
...
+            + "and displays them as a list with index numbers. Search is case-insensitive.\n"
+            + "Parameters: [n/NAME] [nric/NRIC] [role/ROLE] [t/TAG]...\n"
+            + "Example: " + COMMAND_WORD + " n/Alex nric/S6283947C role/patient t/friend";
```
Comment (Ponharish) [inline src/main/java/seedu/address/logic/commands/FindCommand.java:21]: Am I right to assume that not all of the parameters such as name, nric, role, tags are required?
In such a case would it be better to state that not all parameters (such as name, nric, role, ...) are required by the function?

File: src/test/java/seedu/address/model/person/ContainsKeywordsPredicateTest.java [hunk 4]
```diff
@@ -0,0 +1,153 @@
...
+
+        ContainsKeywordsPredicate predicate = new ContainsKeywordsPredicate(mapForKeywords);
+
+        String expected = ContainsKeywordsPredicate.class.getCanonicalName()
+                + "{role/=" + roles + "}";
+
+        assertEquals(expected, predicate.toString());
+    }
+
+}
```
Comment (Ponharish) [inline src/test/java/seedu/address/model/person/ContainsKeywordsPredicateTest.java:153]: So far everything looks good to me. You just got to cover a few more methods with test cases and you would be done.

File: src/test/java/seedu/address/logic/parser/FindCommandParserTest.java
Comment (lavanyagarg112) [inline src/test/java/seedu/address/logic/parser/FindCommandParserTest.java:1]: Nice tests!

**Review by lavanyagarg112 — APPROVED** [review]
LGTM

**Pull request merged**